warehouse/integrations/deltalake-native/deltalake_native.go

```diff
@@ -0,0 +1,1517 @@
+				warehouseutils.UsersTable:      fmt.Errorf("getting partition query: %w", err),
+			}
+		}
+
+		query = mergeableLTSQLStatement(
+			d.Namespace,
+			warehouseutils.UsersTable,
+			stagingTableName,
+			columnKeys,
+			partitionQuery,
+		)
+		queryType = model.Merge
+	}
+
+	startedAt = time.Now()
+
+	row := d.DB.QueryRow(query)
+
+	d.slowQueryLog(queryType, time.Since(startedAt),
+		logfield.TableName, warehouseutils.UsersTable,
+		logfield.Query, query,
+	)
+
+	if row.Err() != nil {
+		return map[string]error{
+			warehouseutils.IdentifiesTable: nil,
+			warehouseutils.UsersTable:      fmt.Errorf("running deduplication: %w", row.Err()),
+		}
+	}
+
+	var (
+		affected int64
+		updated  int64
+		deleted  int64
+		inserted int64
+	)
+
+	if d.LoadTableStrategy == appendMode {
+		err = row.Scan(&affected, &inserted)
+	} else {
+		err = row.Scan(&affected, &updated, &deleted, &inserted)
+	}
+
+	if err != nil {
+		return map[string]error{
+			warehouseutils.IdentifiesTable: nil,
+			warehouseutils.UsersTable:      fmt.Errorf("getting rows affected for dedup: %w", err),
+		}
+	}
+
+	d.Stats.NewTaggedStat("dedup_rows", stats.CountType, stats.Tags{
+		"sourceID":    d.Warehouse.Source.ID,
+		"sourceType":  d.Warehouse.Source.SourceDefinition.Name,
+		"destID":      d.Warehouse.Destination.ID,
+		"destType":    d.Warehouse.Destination.DestinationDefinition.Name,
+		"workspaceId": d.Warehouse.WorkspaceID,
+		"namespace":   d.Namespace,
+		"tableName":   warehouseutils.UsersTable,
+	}).Count(int(updated))
+
+	d.Logger.Infow("completed loading for users and identifies tables",
+		logfield.SourceID, d.Warehouse.Source.ID,
+		logfield.SourceType, d.Warehouse.Source.SourceDefinition.Name,
+		logfield.DestinationID, d.Warehouse.Destination.ID,
+		logfield.DestinationType, d.Warehouse.Destination.DestinationDefinition.Name,
+		logfield.WorkspaceID, d.Warehouse.WorkspaceID,
+		logfield.Namespace, d.Namespace,
+	)
+
+	return map[string]error{
+		warehouseutils.IdentifiesTable: nil,
+		warehouseutils.UsersTable:      nil,
+	}
+}
+
+func (d *Deltalake) tableLocationQuery(tableName string) string {
+	var (
+		enableExternalLocation = warehouseutils.GetConfigValueBoolString(EnableExternalLocation, d.Warehouse)
+		externalLocation       = warehouseutils.GetConfigValue(ExternalLocation, d.Warehouse)
+	)
+
+	if enableExternalLocation != "true" {
+		return ""
+	}
+	if externalLocation == "" {
+		return ""
+	}
+	return fmt.Sprintf("LOCATION '%s/%s/%s'", externalLocation, d.Namespace, tableName)
+}
+
+func (d *Deltalake) dropDanglingStagingTables() {
+	var (
+		tableNames          []string
+		filteredTablesNames []string
+	)
+
+	tableNames, _ = d.fetchTables()
+	for _, tableName := range tableNames {
+		if !strings.HasPrefix(tableName, warehouseutils.StagingTablePrefix(provider)) {
+			continue
+		}
+		filteredTablesNames = append(filteredTablesNames, tableName)
+	}
+
+	d.dropStagingTables(filteredTablesNames)
+}
+
+func Connect(cred Credentials) (*sql.DB, error) {
```

Comment:
I don't see `Connect` being used anywhere else outside of this pkg apart from test. So, won't it make more sense to create a separate test file for non-exported func and then test `Connect` there instead?

Reply:
We are calling this from deltalake_native_test.

Reply:
yup, but that is a test... right? Exporting something just for the sake of testing it, sounds weird to me. Alternative to this, can't we move tests of all unexported functions to maybe a separate file which is in `datalake_native` pkg itself.

Reply:
+1 for not exporting functions that are not meant to be used from outside. in my opinion not exporting unnecessarily symbols limits the API surface that we expose, reducing cognitive load and easing software maintenance. tests in go can use unexported symbols, there is no need to change the package name to have a test being called "integration test".

Reply:
Will take this as a separate task to do this.

Reply:
Great, thanks @achettyiitr, please link the notion task here for posterity :+1:

Reply:
https://www.notion.so/rudderstacks/For-error-mappings-test-that-errors-are-mapped-with-their-corresponding-mapping-0d2740508c2b4eeba80d39a47f20a1f9?pvs=4

Reply:
https://www.notion.so/rudderstacks/Export-functions-for-warehouse-1e1b9177cd3a42c4a2d18fcc2d73f138?pvs=4

Reply:
This has been fixed in the current version.